common/quotas/global/doc.go

```diff
@@ -0,0 +1,45 @@
+// The MIT License (MIT)
+
+// Copyright (c) 2017-2020 Uber Technologies Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+/*
+Package global will contain a global-load-balance-aware ratelimiter.
+
+At a very high level, this will:
+  - collect usage metrics in-memory in "limiting" hosts (which use an in-memory ratelimiter)
+  - asynchronously submit these usage metrics to "aggregating" hosts, which will return per-key RPS data
```

Comment:
do we need to define APIs for submitting the usage metrics?

Reply:
at the RPC level, yes - that'll be in a later PR.

my current plan is to put these in History because (only) Frontend doesn't have "send to host in ring" set up.  in principle it doesn't matter where it is though, and I could add a Frontend ring if we think that's worth it, to keep data within the same kinds of services.

Reply:
history hosts for persisting frontend ratelimit usage data.  (I just added a lot more to both package docs)

"history storing frontend's use" is legitimately a bit weird, but frontend does not currently have ringpop request-forwarding set up so I just grabbed something else because it doesn't actually matter what hosts store the data.

If we would prefer to add a ring for the frontends though, I think we might find other uses for it, and it would be _very_ simple to change from history to frontend.

Reply:
I don't personally have any real concerns with using History as a an in-memory datastore for this, that seems fine. I think elsewhere the naming of one or two variables confused me

Reply:
this is referring to the `history` struct, for tracking "history of request weight"?
you marked that as resolved so I'm assuming it's no longer a concern, but I do agree that "history" is a special term - I'm game to come up with something else if you think it'd help prevent misunderstandings.

Reply:
I'd consider adding a prefix to the struct, but htat's not super consequential, the diagram has helped. 

Regarding request-forwarding: Using history as a means of in-memory aggregation seems perfectly reasonable, but I don't think you need request forwarding (also, I am not sure it's available in the uns implementation?)

Reply:
we need all hosts receiving requests for key X to meet in one place to exchange data, since there's no other way to figure out load distribution.  I'm not sure if that fits with "request forwarding" or not, but that's all it is.

hashing alone is fine to achieve that, but the consistent hash ring means less data loss on host-collection changes (i.e. not "almost everything disappears immediately").  and the UNS implementation supports this just fine, else e.g. matching would be unable to function for handling workflow-query requests.

Reply:
re-did the diagram btw, how does this one look?